### PR TITLE
Support unpacking python dictionary in torch.jit.trace()

### PR DESCRIPTION
**Pull request by tangleintel**
# Support unpacking python dictionary in **torch.jit.trace()**

### Problem 1(usability):
Say, if you have a model and its forward method defined as follows:
**`def forward(self, key1=value1, key2=value2, key3=value3)`**
And you have a dataset and each data point in the dataset is a python dict as follows:
**`data = {key1:value1, key3:value3, key2:value2}`**

The problem is that if you want to trace the model using the dict data by the giving dataset, you need unpack the dictionary and reorder its value manually and make up a tuple as **`data_tuple = (value1, value2, value3)`** as the **`example_inputs`** parameter of **`torch.jit.trace()`**. This marshalling process is not user friendly.

### Problem 2 (feasibility):
Say, if you have a model and its forward method defined as follows:
**`def forward(self, key1=None, key2=None, key3=None)`** -> The default value is **None**
And you have a dataset and each data point in the dataset is a python dict as follows:
**`data = {key1:value1, key3:value3}`** -> Only **part of** the required value by forward was given, the rest use the default value.

The problem is that if you want to trace the model using the dict data by the giving dataset, it's not feasible at all. Cause neither you can pass a tuple like **`T1 = (value1, value3)`**  nor **`T2 = (value1, None, value3)`**. T1 will mismatch value3 with key2 and T2 include **None** type which will be blocked by tracer's type checking. (Of course you can pass **`T3 = (value1,)`**  to make the trace function finish without exception, but the traced model you get probably is not what you expect cause the different input may result in different traced result.).

These problems come from the HuggingFace's PT model, especially in text-classification tasks with datasets such as [MRPC,](https://paperswithcode.com/dataset/mrpc)  [MNLI](https://paperswithcode.com/dataset/multinli) etc.

## Solution
To address these two issues, we propose to support a new type, that is, python dict as example_inputs parameter for torch.jit.trace(). We can base on the runtime type information of the example_inputs object to determine if we fall back to the original tuple path or go into the new dictionary path. Both problem 1 and  problem 2 can be solved by utilizing the "**`**`**" 
operator.

## Limitation & Mitigation

1. If we use dict as example_inputs to trace the model, then we have to pass a dictionary to the traced model too. (Cause probably we will change the order of debug name of the input parameter in torchscript IR, thus we can't assume the traced model's input parameters order are the same with the original model.). We need highlight this too in the document to mitigate this problem.

    For example:
```
# fetch a data from dataloader, and the data is a dictionary
# and the example_inputs_dict is like: {key1:value1, key3:value3, key2:value2}
# the forward() is like: def forward(self, key1=value1, key2=value2, key3=value3)
example_inputs_dict = next(iter(dataloader))
jit_model = model.eval()
# use the dictionary to trace the model
jit_model = torch.jit.trace(jit_model, example_inputs_dict, strict=False)  # Now the IR will be graph(%self : __torch__.module.___torch_mangle_n.Mymodule, %key1 : type1, %key3 : type3, %key2 : type2)
jit_model = torch.jit.freeze(jit_model)

# It's OK to use dict as the parameter for traced model
jit_model(**example_inputs_dict)

example_inputs_tuple = (value1, value3, value2)
# It's wrong to rely on the original args order.
jit_model(*example_inputs_tuple)

```
## Note
1. This PR will make some UT introduced in [39601](https://github.com/pytorch/pytorch/pull/39601) fail, which I think should be classified as unpacking a tuple containing a single dictionary element in our solution.
4. I think there is ambiguity since currently we only specify passing a tuple or a single Tensor as our example_inputs parameter in **torch.jit.trace()**'s documentation, but it seems we can still passing a dictionary.

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel @mingfeima @XiaobingSuper @ashokei @jingxu10

### CODE REVIEW

**Comment by tangleintel** [comment]
Hi,
@davidberard98 
@jamesr66a 
@wanchaol 
@xw285cornell 
@gmagogsfm 
@suo 
@zdevito 
Could you pls help review this PR? Thx!

**Comment by tangleintel** [comment]
> Hi, thank you for the PR! I think traditionally we haven't done this because if you change the keys it might silently succeed without error. Would it be possible to add something similar ? Could you add a test ?
> 
> Will do more thorough review tomorrow.

Hi, eellison
    Thanks for your replay! Can you be more specific on "change the keys" and the correspond case that will silently succeed without error? If you mean change the keys name, I think tracer will report error msg suggesting the name not matching with the arguments name since we can capture and record forwards arguments name. If you mean change the keys (serialization)order, I am not sure if it is the same situation which I noted in the Limitation & Mitigation section.
    I have added one [UT](https://github.com/pytorch/pytorch/pull/81623/files#diff-c54c3bae2ea206c4de5b90adf7f2d3ec47df6818451b8dd2b3c36fe54adc10d8R3031-R3047) which correspond the problem 2 I mentioned above. I can add more if you think it's not informative.

**Comment by tangleintel** [comment]
Hi, @eellison @davidberard98 
    Any comments about this PR?

**Comment by eellison** [comment]
Hi, do you mind adding a test like the following ?

```
def foo(x, y):
    return x + y

out = torch.jit.trace({'x': torch.rand([2]), 'y': torch.rand([2])})
# following should fail
out({'z': torch.rand([2]), 'z': torch.rand([2])}


```

**Comment by tangleintel** [comment]
> Hi, do you mind adding a test like the following ?
> 
> ```
> def foo(x, y):
>     return x + y
> 
> out = torch.jit.trace({'x': torch.rand([2]), 'y': torch.rand([2])})
> # following should fail
> out({'z': torch.rand([2]), 'z': torch.rand([2])}
> ```

@eellison 
Sure, I have added it.([UT](https://github.com/pytorch/pytorch/pull/81623/files#diff-c54c3bae2ea206c4de5b90adf7f2d3ec47df6818451b8dd2b3c36fe54adc10d8R3053-R3059))

**Comment by tangleintel** [comment]
Hi, @eellison 

```
def test_dictionary_as_example_inputs_for_jit_trace(self):
        class TestModule_v1(torch.nn.Module):
            def __init__(self):
                super(TestModule_v1, self).__init__()

            def forward(self, key2=None, key3=None, key4=None, key5=None, key1=None, key6=None):
                return key1 + key2 + key3

        class TestModule_v2(torch.nn.Module):
            def __init__(self):
                super(TestModule_v2, self).__init__()

            def forward(self, x, y):
                return x + y

        model_1 = TestModule_v1()
        model_2 = TestModule_v2()
        value1 = torch.ones(1)
        value2 = torch.ones(1)
        value3 = torch.ones(1)
        example_input_dict = {'key1': value1, 'key2': value2, 'key3': value3}
        traced_model_1 = torch.jit.trace(model_1, example_input_dict, strict=False)
        traced_model_2 = torch.jit.trace(model_2, {'x': torch.rand([2]), 'y': torch.rand([2])})
        res_1 = traced_model_1(**example_input_dict)
        self.assertEqual(res_1, 3 * torch.ones(1)) # 
        with self.assertRaisesRegex(RuntimeError, "forward\(\) is missing value for argument 'x'."):
            res_2 = traced_model_2(**{'z': torch.rand([2]), 'y': torch.rand([2])})
        with self.assertRaisesRegex(RuntimeError, "forward\(\) is missing value for argument 'y'."):
            res_2 = traced_model_2(**{'x': torch.rand([2]), 'z': torch.rand([2])})
```
Above code example is the complete unit test I add for this feature. TestModule_v1 class wants to show the positive example while the TestModule_v2 wants to show the negative example which I think is what you mean right?

**Comment by tangleintel** [comment]
Hi, @eellison 
Sorry for the notification again. May I know when we can review the UT and the PR?

File: test/jit/test_tracer.py
Comment (davidberard98) [inline test/jit/test_tracer.py:2362]: Can we avoid breaking this type of use case? This same usage would break for other users with similar usage patterns (i.e. `trace(model, example_inputs={'forward': dict})`).

Instead, can we provide a flag that disables the `dict` -> `tuple(dict)` conversion? e.g. `torch.jit.trace(model, inputs, unpack_input_dict=True)`. And `unpack_input_dict` would default to `False`.

File: torch/csrc/jit/python/script_init.cpp
Comment (davidberard98) [inline torch/csrc/jit/python/script_init.cpp:1186]: Could we leave this with its old name? Even though this method is prefixed with an underscore, I would not be surprised if there are users who use this api directly...

File: torch/csrc/jit/python/python_tracer.cpp
Comment (davidberard98) [inline torch/csrc/jit/python/python_tracer.cpp:105]: Could we reorder the inputs instead of the argument names? This might make more sense to people when they trace a model, so that their arguments don't get reordered compared to the method definition.

File: torch/jit/_trace.py
Comment (davidberard98) [inline torch/jit/_trace.py:981]: also commented above, but I think it would be better if this behavior was guarded by an additional parameter passed into torch.jit.trace.

Comment (davidberard98) [reply torch/jit/_trace.py:981]: also, please add documentation of this flag to the documentation for trace_method.

And, if you want to expose this flag in torch.jit.trace(), I think it would be preferable to also support unpacking python dicts in torch.jit.trace for functions as well. But if you don't want to add that I think it's fine, in that case don't add the flag to torch.jit.trace.

File: test/test_jit.py
Comment (davidberard98) [inline test/test_jit.py:3050]: just curious, how does it pick up on the fact that key4, key5, key6 are not needed? Is it based on the presence of a default value?

**Review by davidberard98 — CHANGES_REQUESTED** [review]
Hey @tangleintel , sorry this has taken so long to review. I appreciate the contribution and this does look like a very helpful feature!

I added a few requests inline. In particular I think this will break some use cases (the ones that you already had to patch in the tests). I think it would be better if we guard this behavior with an extra parameter to avoid breaking existing use cases.

**Comment by davidberard98** [comment]
also there's a failing lint, please fix that (I know you couldn't see this earlier because the check required an approval to run)

File: test/jit/test_tracer.py
Comment (tangleintel) [reply test/jit/test_tracer.py:2362]: Hi, @davidberard98 
Thanks for your review!
For this use case, I personally think it is not appropriate. My protest opinion comes from below two point:
1. The current PyTorch documentation for [torch.jit.trace()](https://pytorch.org/docs/stable/generated/torch.jit.trace.html) specifies that the **example_inputs** parameter should come with the type of _tuple_ and _torch.Tensor_. I'm not sure these two type for **example_inputs** argument is a must or just a suggestion. If it is a must, this usage case shouldn't been included at first. At least, this undocumented behavior shouldn't been view as a standard.
2. Our solution is for the problem of how to pass variable args(positional) and kwargs in torch.jit.trace() for a given module. In this respective, a tuple is used for a packed of positional args while dict is used for pass keyword arguments, and dict used by this example should be viewed as an element of tuple since tracer support tuple with nested collection types. So, our dict is not in the same dimension with the this one.

For your suggestion with an additional argument to address backward compatibility:
1. I don't know if it is important to maintain this undocumented behavior for PyTorch with such a flag.
2. With this design of API, we need educate the user that the dict would substantially comes with two different meanings and the semantic of **example_inputs** parameter should depend on another one, that is, **unpack_input_dict**. For my opinion, making the semantics of different parameters as orthogonal as possible will give users a cleaner API.
3.  Inside the [tracer](https://github.com/pytorch/pytorch/pull/81623/files#diff-09665607e34394eac04617a39929e7ca9b732af62eb39dad48b1e47b733b2885L341), we have to deal with the case or a single tensor and dict. If we give user a unified front end interface, maybe we don't have to deal with such corner case.

Anyway, above is just my opinion and I think the decision of how to design the interface is up to PyTorch team.

File: torch/csrc/jit/python/script_init.cpp
Comment (tangleintel) [reply torch/csrc/jit/python/script_init.cpp:1186]: OK, this old name could be kept.(Pls. I think this is the same issue about if we should maintain the backward compatibility for a non-standard behavior right?)

File: torch/jit/_trace.py
Comment (tangleintel) [reply torch/jit/_trace.py:981]: Yes, once we reach an agreement with the API design, I will do the corresponding modification and update the document.

File: test/test_jit.py
Comment (tangleintel) [reply test/test_jit.py:3050]: Yes, it is

**Comment by tangleintel** [comment]
> also there's a failing lint, please fix that (I know you couldn't see this earlier because the check required an approval to run)

Sure.

File: test/jit/test_tracer.py
Comment (tangleintel) [reply test/jit/test_tracer.py:2362]: Hi, @davidberard98 
Thanks for your review!
For this use case, I personally think it is not appropriate. My protest opinion comes from below two points:
1. The current PyTorch documentation for [torch.jit.trace()](https://pytorch.org/docs/stable/generated/torch.jit.trace.html) specifies that the **example_inputs** parameter should come with the type of _tuple_ or _torch.Tensor_. I'm not sure these two type for **example_inputs** argument is a must or just a suggestion. If it is a must, this usage case shouldn't been included at first. At least, this undocumented behavior shouldn't been view as a standard.
2. Our solution is for the problem of how to pass variable args(positional) and kwargs in torch.jit.trace() for a given module. In this respective, a tuple is used for a packed of positional args while dict is used for pass keyword arguments, and dict used by this example should be viewed as an element of tuple since tracer support tuple with nested collection types. So, our dict is not in the same dimension with the this one.

For your suggestion with an additional argument to address backward compatibility:
1. I don't know if it is important to maintain this undocumented behavior for PyTorch with such a flag.
2. With this design of API, we need educate the user that the dict would substantially comes with two different meanings and the semantic of **example_inputs** parameter should depend on another one, that is, **unpack_input_dict**. For my opinion, making the semantics of different parameters as orthogonal as possible will give users a cleaner API.
3.  Inside the [tracer](https://github.com/pytorch/pytorch/pull/81623/files#diff-09665607e34394eac04617a39929e7ca9b732af62eb39dad48b1e47b733b2885L341), we have to deal with the case or a single tensor and dict. If we give user a unified front end interface, maybe we don't have to deal with such corner case.

Anyway, above is just my opinion and I think the decision of how to design the interface is up to PyTorch team.

File: test/test_jit.py
Comment (tangleintel) [reply test/test_jit.py:3050]: Yes, it is

File: torch/csrc/jit/python/python_tracer.cpp
Comment (tangleintel) [reply torch/csrc/jit/python/python_tracer.cpp:105]: It seems feasible. I will try it locally, and if I verify the result, I will follow your suggestion.

File: torch/jit/_trace.py
Comment (tangleintel) [reply torch/jit/_trace.py:981]: Yes, once we reach an agreement with the API design, I will do the corresponding modification and update the document.

File: torch/csrc/jit/python/python_tracer.cpp
Comment (tangleintel) [reply torch/csrc/jit/python/python_tracer.cpp:105]: Hi, @davidberard98 
I think we have to do this reordering. 
According to this [code snippet](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/frontend/tracer.cpp#L490-L508):
```
// When enough argument name hints are provided, use them as debug names
// for traced function/modules.
// Here argument_names is allowed to have more names than needed because
// some arguments may have valid default values, therefore they don't need
// example inputs.
if (argument_names.size() >= inputs.size()) {
  for (size_t i = 0, e = inputs.size(); i < e; ++i) {
    IValue& input = inputs[i];
    input = addInput(
        state,
        input,
        input.type(),
        state->graph->addInput(argument_names[i]));
  }
} else {
  for (IValue& input : inputs) {
    input = addInput(state, input, input.type(), state->graph->addInput());
  }
}
```
Pytorch will set the debug name of the traced torchscript according to the real input's order, in fact, I think what matters is the number of the real input other than the order. Use the example I used in the UT and RFC's Problem 2 in the first section, there was missing arguments, which cause  
the input and argments_name not match with each other. So, one of the intention here is to compact the argument_name and make them align with inputs in length.(I think it's impossible(at least I haven't come out a decent method) to expand the input with the default value of the missing args.), and another is reorder.
The debug name will be used in [_check_trace()](https://github.com/pytorch/pytorch/blob/master/torch/jit/_trace.py#L310) after the tracing procedure.

**Comment by tangleintel** [comment]
Hi, @davidberard98 
I have followed your suggestion to add an additional flag in torch.jit.trace() to suggest the additional meaning of python dictionary as example_input in the latest patch, please help review it. If you think it is not right, we can revert it anytime.
Pls, can you help me to trigger the CI since I haven't the permission.

File: test/jit/test_tracer.py
Comment (davidberard98) [reply test/jit/test_tracer.py]: @tangleintel I definitely see your point; however I discussed with some other pytorch maintainers who agree with me - i.e. even though this isn't documented, torch.jit.trace is widely used and we'd prefer to avoid breaking use cases that depend on the current behavior.

We can probably deprecate the old behavior with dict inputs, e.g. see [the 180 days + 2 releases policy](https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy). However I'd probably need your help pushing an update after 180 days, since I'd probably forget :)

File: torch/csrc/jit/python/python_tracer.cpp
Comment (davidberard98) [reply torch/csrc/jit/python/python_tracer.cpp:105]: Hmm, I think I understand... Can you do something like this instead?

```c++
for (i in 0..argument_names.size()) {
  if (argument_names[i] in input_dict) {
    reordered_argument_names.push_back(argument_names[i])
  }
}
```

Or am I misunderstanding the problem?

**Review by davidberard98 — COMMENTED** [review]
High-level question (suggested by another developer at Meta): would something like this work?

```
torch.jit.trace_module(my_mod, {"forward": torch.jit.make_arguments_tuple(my_mod.forward, my_input_dict)})
```

(where `make_arguments_tuple` would need to be implemented, e.g. using `inspect` to identify the arguments to my_mod.forward). This would have slightly different semantics compared to this PR, I haven't thought too much about the pros and cons - but it would skip questions about ordering/etc.

Also, you might need to rebase & check the lints again.

**Comment by tangleintel** [comment]
> High-level question (suggested by another developer at Meta): would something like this work?
> 
> ```
> torch.jit.trace_module(my_mod, {"forward": torch.jit.make_arguments_tuple(my_mod.forward, my_input_dict)})
> ```
> 
> (where `make_arguments_tuple` would need to be implemented, e.g. using `inspect` to identify the arguments to my_mod.forward). This would have slightly different semantics compared to this PR, I haven't thought too much about the pros and cons - but it would skip questions about ordering/etc.
> 
> Also, you might need to rebase & check the lints again.

Hi, @davidberard98 
I think this indeed can resolve the ordering issue, but the problem(parameter's number not matching issue) still comes when one of the missing argument's default value is None. There are two situations when some of the argument is missing:
1. When the missing argument's default value is not None, users can insert the default value to the missing position in `my_input_dict` **manually**(I don't know if there is a way to fill the missing args **automatically**, as far as I known, [`inspect.signature`](https://docs.python.org/3/library/inspect.html#introspecting-callables-with-the-signature-object) can only parse the function's signature literally, so, we can get the string of the arguments' name and the "string" of the default value, but the real type object(I think maybe it's the dynamic type nature of python makes this difficult)).
2. When the missing argument's default value is None. We can't make a tuple such as `a_tuple = (value1, value2, None, value4)` passing to tracer, cause tracer will block it and throw an exception when one of the element of tuple is `None`. 
This is why we do this ugly things inside tracer, otherwise, we can just to the reordering things before entering tracer as you mentioned.

File: torch/csrc/jit/python/python_tracer.cpp
Comment (tangleintel) [reply torch/csrc/jit/python/python_tracer.cpp:105]: [Commented below](https://github.com/pytorch/pytorch/pull/81623#issuecomment-1269542950)

File: test/jit/test_tracer.py
Comment (tangleintel) [reply test/jit/test_tracer.py]: @davidberard98 
OK, then let's follow your suggestion to maintain the backward compatibility and follow the Pytorch's BC-breaking rules to mark the undocumented behavior as deprecated. Is it OK for you?
BTW, according to the BC-breaking rules, if we want to deprecate an old behavior, we need specify which pytorch version the deprecation first takes effect in a warning msg. So, which version you think this feature can be landed in(and the deprecated in effect version is at least plus 2 right?) ?
 Yes, we will consistently track this feature and give the update when time is right.

**Comment by tangleintel** [comment]
Sorry, I haven't fixed the lint error when I push the updated commit and the CI is automatically triggered.

**Comment by davidberard98** [comment]
@tangleintel I still see some lint failures, I think you need to rebase to fix those.

Also I added an inline comment above that I think you may have missed.

**Comment by tangleintel** [comment]
Hi, all
Sorry to involve you in this PR(I think it's my mistake in rebase process that result you are involved).

File: torch/csrc/jit/python/python_tracer.cpp
Comment (tangleintel) [reply torch/csrc/jit/python/python_tracer.cpp:105]: Hi, @davidberard98 
Do you mean put this logic in python other than C++?

**Comment by justinchuby** [comment]
cc @shubhambhokare1 @BowenBao

File: torch/csrc/jit/python/python_tracer.cpp
Comment (tangleintel) [reply torch/csrc/jit/python/python_tracer.cpp]: Hi, @davidberard98 
This way will not reorder the arguments, just compact. I followed your suggestions but this need modify the `traced_input` too.

**Comment by tangleintel** [comment]
> @tangleintel I still see some lint failures, I think you need to rebase to fix those.
> 
> Also I added an inline comment above that I think you may have missed.

Hi, @davidberard98 
I have rebased to master. The lint error should comes from a missing file(pr-sanity-check.sh) in my feature branch.
[This](https://github.com/pytorch/pytorch/pull/81623#discussion_r988224496) is the missing comment right? I have followed your suggestions, you may not see the inline comment but can view my reply above.

**Review by davidberard98 — COMMENTED** [review]
@tangleintel a few comments about the interface:

- After discussing with some other developers they have commented that they prefer not to add any BC-breaking changes, even if follow the deprecated-for-180-days process, because this process is reserved for fixing incorrect behavior, and this doesn't really fit into that category.
- They recommend an interface as follows: provide either `example_inputs` which behaves exactly how it does now, or provide `example_kwarg_inputs`, which has the dict-unpacking behavior that you want. i.e. "explicit is better than implicit"
- However I'm okay with leaving the interface as is if you would prefer to stop making changes to this PR and just get it landed.

**Comment by tangleintel** [comment]
> @tangleintel a few comments about the interface:
> 
> * After discussing with some other developers they have commented that they prefer not to add any BC-breaking changes, even if follow the deprecated-for-180-days process, because this process is reserved for fixing incorrect behavior, and this doesn't really fit into that category.
> * They recommend an interface as follows: provide either `example_inputs` which behaves exactly how it does now, or provide `example_kwarg_inputs`, which has the dict-unpacking behavior that you want. i.e. "explicit is better than implicit"
> * However I'm okay with leaving the interface as is if you would prefer to stop making changes to this PR and just get it landed.

Hi, @davidberard98 
I followed your college's situation which I think is reasonable too(explicitly specified while keep the compatibility). But this additional argument may result that the original `example_inputs` argument become a keyword argument and its default value is None, since I think in this situation, either `example_inputs` or 'example_kwarg_inputs` should be specified.

**Comment by tangleintel** [comment]
Hi, @davidberard98 
Can you  pls help me check the `Meta Internal-Only Changes Check` failure in the CI?

File: torch/jit/_trace.py
Comment (davidberard98) [inline torch/jit/_trace.py:613]: my understanding is that torch.jit.trace() will _only_ work with example_kwarg_inputs if the traced object is a module (and not if it's a simple function). Is that true?

If so, can we only introduce this interface to `trace_module` for now? If you want to add support for all objects (including `torch.jit.trace(fn, example_kwarg_inputs=...)` for functions) you can do so in a separate PR if that's easier. But for now I think it would make more sense to only expose this in torch.jit.trace_module, if it only works for modules and not for functions.

File: torch/jit/_trace.py
Comment (davidberard98) [inline torch/jit/_trace.py:671]: (related to the other comment above) Can you move these changes to the documentation into `def trace_module` instead of `def trace`?

(Or, if this is already supporting `torch.jit.trace(function, example_kwarg_inputs=...)`, then it would be good anyway to have the `example_kwarg_inputs` option documented in both trace and trace_module).

**Review by davidberard98 — COMMENTED** [review]
@tangleintel hey sorry, one other request regarding documentation & interface

File: torch/jit/_trace.py
Comment (tangleintel) [reply torch/jit/_trace.py:613]: Hi, @davidberard98 
Thanks for you remainder! I did just support **module** trace in torch,jit.trace(). But I think for user it is more convenient to use torch.jit.trace() to trace a module(at least for me, I don't have to wrap the example input again(such as ``{'forward': example_inputs}``)). So I decide to complete the  
support for function in torch.jit.trace() and add the doc for ``trace_module()``.

File: torch/jit/_trace.py
Comment (tangleintel) [reply torch/jit/_trace.py:671]: As commented above.

**Comment by davidberard98** [comment]
@tangleintel please address the one remaining lint.. and then I think this should be good to go!

**Comment by tangleintel** [comment]
rebase again.

**Comment by tangleintel** [comment]
@davidberard98 
Sorry for these annoying lint issues again and again.(It seems there is some difference between my flake8 and pytorch's lint tool.)

**Comment by tangleintel** [comment]
Hi, @davidberard98 
I think the failed test("linux-focal-py3.7-clang-onnx") is introduced by this commit id ``960b98128e`` after I rebased on master.

**Comment by tangleintel** [comment]
rebase again

**Comment by davidberard98** [comment]
@tangleintel fyi in case you have this issue again with broken tests on master, you can instead rebase on top of the viable/strict branch, which points to the latest commit for which all the tests have passed.

**Comment by tangleintel** [comment]
> @tangleintel fyi in case you have this issue again with broken tests on master, you can instead rebase on top of the viable/strict branch, which points to the latest commit for which all the tests have passed.

Thx @davidberard98 ! I just rebased.

**Review by davidberard98 — APPROVED** [review]
looks good!

**Pull request closed without merging**
